packages/pricing/src/services/pricing-module.ts

```diff
@@ -248,12 +281,247 @@ export default class PricingModuleService<
+      return acc
+    }, new Map())
+
+    const invalidRuleAttributes = ruleAttributes.filter(
+      (r) => !ruleTypeMap.has(r)
+    )
+
+    if (invalidRuleAttributes.length > 0) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Rule types don't exist for: ${invalidRuleAttributes.join(", ")}`
+      )
+    }
+
+    const invalidMoneyAmountRule = data
+      .map((d) => d.prices?.map((ma) => Object.keys(ma.rules)).flat() ?? [])
+      .flat()
+      .filter((r) => !ruleTypeMap.has(r))
+
+    if (invalidMoneyAmountRule.length > 0) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Rule types don't exist for money amounts with rule attribute: ${invalidMoneyAmountRule.join(
+          ", "
+        )}`
+      )
+    }
+
+    const priceSets = await Promise.all(
+      data.map(async (d) => {
+        const { rules, prices, ...rest } = d
+        const [priceSet] = await this.priceSetService_.create(
+          [rest],
+          sharedContext
+        )
+
+        if (rules?.length) {
+          const priceSetRuleTypesCreate = rules!.map((r) => ({
+            rule_type: ruleTypeMap.get(r.rule_attribute),
+            price_set: priceSet,
+          }))
+
+          await this.priceSetRuleTypeService_.create(
+            priceSetRuleTypesCreate as unknown as PricingTypes.CreatePriceSetRuleTypeDTO[],
+            sharedContext
+          )
+        }
+
+        if (prices?.length) {
+          for (const ma of prices) {
+            const [moneyAmount] = await this.moneyAmountService_.create(
+              [ma] as unknown as CreateMoneyAmountDTO[],
+              sharedContext
+            )
+
+            const numberOfRules = ma.rules ? Object.entries(ma.rules).length : 0
+
+            const [priceSetMoneyAmount] =
+              await this.priceSetMoneyAmountService_.create(
+                [
+                  {
+                    price_set: priceSet,
+                    money_amount: moneyAmount,
+                    title: "test",
+                    number_rules: numberOfRules,
+                  },
+                ] as unknown as PricingTypes.CreatePriceSetMoneyAmountDTO[],
+                sharedContext
+              )
+
+            if (numberOfRules) {
+              const priceSetRulesCreate = Object.entries(ma.rules).map(
+                ([k, v]) => ({
+                  price_set_money_amount: priceSetMoneyAmount,
+                  rule_type: ruleTypeMap.get(k),
+                  price_set: priceSet,
+                  value: v,
+                  price_list_id: "test",
+                })
+              )
+
+              await this.priceRuleService_.create(
+                priceSetRulesCreate as unknown as PricingTypes.CreatePriceRuleDTO[],
+                sharedContext
+              )
+            }
+          }
+        }
+
+        return priceSet
+      })
+    )
+
+    return priceSets
+  }
+
+  async addRules<
```

Comment:
Todo: I think that it would be best to declare the overloading in the service explicitly and then the implemented method to improve the DX. As you can see in the screen shots, the first one is the intelisense tat dev can receive when the overloading is provided (if the service is typed as the service itself and not the interface) and the second ss correspond to the doc the dev receive without the overloading. Wdyt guys?

<img width="975" alt="Screenshot 2023-10-02 at 14 23 38" src="https://github.com/medusajs/medusa/assets/25098370/da3f6d2b-ce6e-41bb-ac63-25eff9b036a2">
<img width="998" alt="Screenshot 2023-10-02 at 14 24 25" src="https://github.com/medusajs/medusa/assets/25098370/cddacb3b-022c-482c-ae05-b71b5ec0a774">

Reply:
depending on your choice we can decide to always apply this approach in the future and in that service.

Reply:
should be fixed, one note though, we should probably type these services by the interface `IPricingModuleService` rather than the direct implementation from the service, I've updated the service so you get the same typing there 😄

Reply:
I agree, but it is to be consistent in the different places in case the user is just initializing the module and typing it as the service :)